source/simulator/initial_conditions.cc

```diff
@@ -42,6 +45,136 @@ namespace aspect
+        const std::vector<Point<dim> > support_points
+          = finite_element.base_element(base_element).get_unit_support_points();
+        Assert (support_points.size() != 0,
+                ExcInternalError());
+
+        // create an FEValues object with just the temperature/composition element
+        FEValues<dim> fe_values (*mapping, finite_element,
+                                 support_points,
+                                 update_quadrature_points);
+
+        std::vector<types::global_dof_index> local_dof_indices (finite_element.dofs_per_cell);
+
+        for (typename DoFHandler<dim>::active_cell_iterator cell = dof_handler.begin_active();
+             cell != dof_handler.end(); ++cell)
+          if (cell->is_locally_owned())
+            {
+              fe_values.reinit (cell);
+
+              // go through the temperature/composition dofs and set their global values
+              // to the temperature/composition field interpolated at these points
+              for (unsigned int i=0; i<finite_element.base_element(base_element).dofs_per_cell; ++i)
+                {
+                  // if it is specified in the parameter file that the sum of all compositional fields
+                  // must not exceed one, this should be checked
+                  if (parameters.normalized_fields.size()>0 && n == 1)
+                    {
+                      double sum = 0;
+                      for (unsigned int m=0; m<parameters.normalized_fields.size(); ++m)
+                        sum += compositional_initial_conditions->initial_composition(fe_values.quadrature_point(i),parameters.normalized_fields[m]);
+                      if (std::abs(sum) > 1.0+1e-6)
+                        {
+                          max_sum_comp = std::max(sum, max_sum_comp);
+                          normalize_composition = true;
+                        }
+                    }
+
+                }
+            }
+
+        // if at least one processor decides that it needs
+        // to normalize, do the same on all processors.
+        if (Utilities::MPI::max (normalize_composition ? 1 : 0,
+                                 mpi_communicator)
+            == 1)
+          {
+            const double global_max
+              = Utilities::MPI::max (max_sum_comp, mpi_communicator);
+
+            if (n==1)
+              pcout << "Sum of compositional fields is not one, fields will be normalized"
+                    << std::endl;
+
+            for (unsigned int m=0; m<parameters.normalized_fields.size(); ++m)
+              if (n==parameters.normalized_fields[m])
+                initial_solution /= global_max;
+          }
+
+        // then apply constraints and copy the
+        // result into vectors with ghost elements. to do so,
+        // we need the current constraints to be correct for
+        // the current time
+        compute_current_constraints ();
+        current_constraints.distribute(initial_solution);
```

Comment:
I see now that this is how it was already done in the existing code. Will open a new issue to remind ourselves that this could be made more efficient.